Make requests and responses extensible

chrismytton opened the pull request:
This pull request adds several classes that make it easier to hook into the request/response flow and add user-defined extensions at certain points. This uses two concepts: request strategies and response decorators. The request strategies provide a way of saying how to obtain a response in the first place, the response decorators apply a set of transforms to the response, or store the response elsewhere, e.g. in the archive.

# Notes to reviewer

This is quite a big pull request. I think it makes sense to put all of these changes though together though, since they are all related to the goal of making the request/response cycle customizable.

Fixes #10 
Fixes #15 
Fixes #16

chrismytton committed:
Only deal with responses in Scraped class

Rather than having to take a url and a strategy and then combine them to
get a response this simplifies things so that `ScrapedPage` subclasses
only need to deal with responses.

chrismytton committed:
Add prototype of Scraped::Request class

chrismytton committed:
Update README with new interface for Request

tmtmtmtm commented on lib/scraped_page.rb:
```diff
@@ -2,6 +2,7 @@
 require 'scraped_page/version'
 require 'scraped_page/strategy/live_request'
 require 'scraped_page/strategy/live_request_archive'
+require 'scraped_page/request'
```
time to pull in https://github.com/jarmo/require_all?

tmtmtmtm commented on lib/scraped_page/request.rb:
```diff
@@ -0,0 +1,28 @@
+
+    private
+
+    def readers_response(url)
```
perhaps something like `first_successful_response` might be a bit more explicit?

chrismytton committed:
Add require_all gem

This makes it much easier to require all of the code in `lib` that the
gem uses, rather than having to have a require statement for each line.

chrismytton committed:
Rename readers_response -> first_successful_response

This makes it a bit clearer what the method is actually returning. I've
also fixed a bug where `readers_response` was taking an unnecessary
`url` argument.

chrismytton committed:
Move response processors to the response method

This adds a `processors` argument to `Request#response` which can
provide a list of objects that respond to `call` and take a response,
process it and optionally return a new response.

chrismytton committed:
Fix rubocop error with reduce parameter names

This is apparently what Rubocop wants us to name the parameters to a
reduce call.

chrismytton committed:
Return the processed response from request

We want the response _after_ it has been processed, so just use the
return value from the call to `processors.reduce`.

chrismytton committed:
Add a Scraped::Processor class

This acts as an abstract class that processors can inherit from,
allowing the subclasses to just declare methods for the parts of the
response that they want to modify.

chrismytton committed:
Disable the Style/SingleLineBlockParams cop

This is quite an opinionated cop that on Rubocop 0.44.1 wants you to
name your reduce block params `|a, e|` and in Rubocop 0.45.0 it wants
you to name them `|acc, elem|`. So I was getting one failure locally and
a different one on Travis. Disabling this for now, then we can just use
names that make sense.

https://github.com/bbatsov/rubocop/blob/6907b13320493b334dc60a1a7687fc381f1f3740/config/default.yml#L886-L893

chrismytton committed:
Rename reduce block parameters in Request#response

Now we've disabled the cop for this we can give them names that make (a
bit) more sense for our situation.

chrismytton commented:
Following on from discussions in Slack I've added a `ScrapedPage::Processor` class which is intended to be subclassed by things that want to process the response in some way. You can see the new interface in action over in https://github.com/everypolitician-scrapers/spain_congreso_es/pull/15.

chrismytton commented:
@tmtmtmtm :eyes:

tmtmtmtm commented on lib/scraped_page/request.rb:
```diff
@@ -0,0 +1,26 @@
+    end
+
+    def response(processors = [])
+      abort "Failed to fetch #{url}" if first_successful_response.nil?
```
Tiny nit, but `first_successful_response.nil?` caused me a double-take, as it read more like "We got a response, but it was nil" rather than "we didn't get any response". Perhaps `unless first_successful_response` would be clearer?

chrismytton committed:
Remove Scraped::Processor class

This is essentially just a response with a `call` method, so we're
better just using a `Response` and then we can have decorated responses.

chrismytton committed:
Rename Scraped::Request#response parameter

This makes it clearer what we're expecting to have passed to this
method.

chrismytton commented:
Quick update: this is now using `Scraped` rather than `ScrapedPage`, which is definitely nicer. I've removed the `Scraped::Processor` class, which was essentially just acting as a `Response` with a call method.

As discussed in Slack, having response decorators seems like a good approach, but ideally we want to be able to pass some configuration to the constructor, e.g. for the archiver we might want to pass `:repo` and `:branch` options. However the current `Scraped::Response` class expects `:body`, `:status`, `:url`, `:headers` as the arguments to the constructor.

We were working around this with the `Scraped::Processor#call` method, which essentially set an extra instance variable before gathering up the different bits of the response, but that approach seems even more wrong now we're potentially talking about a `Response#call` method.

@tmtmtmtm Any thoughts on how we might be able to have response decorators that can also be configured like a regular class?

tmtmtmtm commented:
@chrismytton without having explored the implementation issues in a lot of depth, what I'm imagining is a base `Scraped::Response::Decorated` class for each of them to inherit from, but which itself _contains_ the `Scraped::Response` (or indeed another `Scraped::Response::Decorated`, matryoshka-style), and delegates any unknown methods to it. Then each of these should be able to have their own configuration options to the constructor.

chrismytton committed:
Merge remote-tracking branch 'origin/master' into add-scraped_page-request-class

chrismytton committed:
Add Scraped::Response::Decorator class

This class encapsulates a response, the configuration for the decorator
and the logic for manipulating a response in some way before it's passed
to the scraper.

chrismytton committed:
Add Scraped::Request::Strategy class

This is the class that should be subclassed in order to provide a custom
request strategy.

chrismytton committed:
Update README documentation

chrismytton committed:
Fix rubocop error

chrismytton commented:
I've pushed some more changes to harmonize the custom requests and the custom responses. They now both follow a similar pattern where the subclass is passed into the array of strategies or decorators. If the user wants to configure the class they can pass in a two-element array where the second element is a hash. This hash is then passed to the constructor as the `config` option, which can then be used in the subclass methods.

You can see the `Scraped::Response::Decorator` class in action in the Spanish scraper in [`absolute_links.rb`](https://github.com/everypolitician-scrapers/spain_congreso_es/blob/b2efca5ec542fa55be6cff7929deb3565b7e9e44/lib/absolute_links.rb) and [`remove_session_information_from_links.rb`](https://github.com/everypolitician-scrapers/spain_congreso_es/blob/scraped_page-request-prototype/lib/remove_session_information_from_links.rb).

N.B. I haven't got a use case for constructor arguments quite yet, so can't links to an example. I should do very shortly when I implement this on a scraper that needs the archiver though.

You can see `Scraped::Request::Strategy` in action in [`Scraped::Request::Strategy::LiveRequest`](https://github.com/everypolitician/scraped/blob/5686b1f822ed623130a3673b1a6677c840e90817/lib/scraped/request/strategy/live_request.rb).

chrismytton commented:
@tmtmtmtm 👀 

Unless there are any glaring errors I'm keen to get this merged tomorrow and get the snowball of scraper change rolling!

tmtmtmtm commented on README.md:
```diff
@@ -43,7 +43,7 @@ end
 Then you can create a new instance and pass in a url
```
This is now out of date. This no longer takes a URL.

tmtmtmtm commented on README.md:
```diff
@@ -55,39 +55,80 @@ page.to_h
+## Extending
 
-The default strategy for retrieving pages is `Scraped::Strategy::LiveRequest`. If you'd also like to archive a copy of the page you're scraping into a git branch, you can pass a `:strategy` option to the constructor:
+There are two main ways to extend `scraped` with your own custom logic - custom requests and decorated responses. Custom requests allow you to change where the scraper is getting its responses from, e.g. you might want to make requests to archive.org if the site you're scraping has disappeared. Decorated responses allow you to manipulate the response before it's passed to the scraper, for example ...
```
comma splice in last sentence

tmtmtmtm commented on README.md:
```diff
@@ -55,39 +55,80 @@ page.to_h
-This will use the [`scraped_page_archive`](https://github.com/everypolitician/scraped_page_archive)
-gem to store a copy of the pages you scrape in a git branch in your scraper's repo.
+The `response` method should return a `Response` instance. You need to pass at least `url` and `body` parameters to the `Response` constructor, you can also pass `status` and `headers` parameters.
+
```
comma splice

tmtmtmtm commented on README.md:
```diff
@@ -55,39 +55,80 @@ page.to_h
 
-A strategy is an object that implements a `response` method which returns a
-`Hash` with at least a `:body` key and optionally `:status` and `:headers` keys.
+To manipulate the response before it is passed to the scraper create a class that subclasses `Scraped::Response::Decorator`:
```
I think this needs a second half that explains up-front what you need to do/provide/return with your subclass, e.g. "… and provides a replacement `body`, `url`, `status`, and/or `headers`")

tmtmtmtm commented on README.md:
````diff
@@ -55,39 +55,80 @@ page.to_h
 ```
 
-Then you can pass that strategy when creating an instance of your `Scraped` subclass:
+As well as the `body` method you can also supply your own `url`, `status` and `headers` methods. You can access the original response instance via the `response` method, as with `response.body` and `response.url` above.
````
What does "original" mean here in the case where there are multiple decorators? Is it actually the original original, or just the one at the point where this class takes over?

tmtmtmtm commented on README.md:
````diff
@@ -55,39 +55,80 @@ page.to_h
+
+```ruby
+request = Scraped::Request.new(url: 'http://example.com')
+response = request.response([AbsoluteLinks])
````
it seems a little odd that this one takes a plain argument, rather than a named parameter.

tmtmtmtm commented on README.md:
````diff
@@ -55,39 +55,80 @@ page.to_h
+response = request.response([AbsoluteLinks])
+```
+
+### Configurating requests and responses
````
"Configuring", I suspect?

tmtmtmtm commented on README.md:
```diff
@@ -55,39 +55,80 @@ page.to_h
+
+response = Scraped::Request.new(url: url).response([
+  [CustomHeader, greeting: 'Hello, world']
+])
```
I'm open to persuasion on this, but this feels quite ugly and kludgy to me. Perhaps better to make this a hash?

```ruby
response = Scraped::Request.new(url: url).response([
   { decorator: CustomHeader, greeting: 'Hello, world' }
 ])
```

tmtmtmtm reviewed (changes requested):
some initial comments purely on the README, before getting around to the code itself.

chrismytton committed:
Use named parameters in Request#response

This is consistent with everywhere else that we're using decorators.

chrismytton committed:
Update README

chrismytton committed:
Configure response decorators with a hash instead of array

Previously we wanted users to configure decorators with a two-element
array:

    [Archive, repo: 'foo/bar', branch: 'baz']

This was very simple to implement internally, but a bit opaque to the
end user looking at the code, especially the weird hash-in-an-array
syntax.

To fix this problem we're switching to instead taking an array for
configuration:

    { decorator: Archive, repo: 'foo/bar', branch: 'baz' }

This has made the _internal_ implementation (slightly) more complicated,
but externally it should be a bit clearer how this should be used.

Ideally all of this interface will be hidden behind some nicer syntax,
but for now we're sticking to the philosophy of explicit is better than
implicit, at least while we're still in the early phases of development.

chrismytton committed:
Fix typo in README

chrismytton committed:
fixup! Configure response decorators with a hash instead of array

chrismytton committed:
Rename readers -> strategies in Request class

This is consistent with what I've documented in the README.

chrismytton committed:
Use hashes rather than arrays to configure strategies

As with 1894b74 it's a bit more obvious what's going on if we configure
strategies with hashes rather than an array.

chrismytton committed:
Add tests for the Response class

chrismytton committed:
Add tests for Scraped::Request class

chrismytton committed:
Add more tests for Scraped::Request class

This is an important class because it handles talking to the request
strategies to get a response and then passing that response through the
response decorators.

chrismytton committed:
Default to empty config in Strategy

chrismytton committed:
Add tests for Scraped::Request::Strategy class

chrismytton committed:
Add tests for Scraped::Response::Decorator class

chrismytton committed:
Update README to use hashes for configuring decorators

chrismytton commented:
@tmtmtmtm 👀

tmtmtmtm reviewed (approved):
There are still a few things to tidy up here (e.g. I think we need to document better what a basic `Scraped::Request `is, as currently we introduce it _en passant_ as how you get a `Scraped::Response` with no real explanation), and it would be good to find a way to remove the duplication of the Class-or-Hash argument shuffle. 

I also suspect we might want a better user-facing name than 'strategies:' but it should be easy enough to do that in a backwardsly compatible manner when we come up with something.

So I think this is definitely good to merge and evolve further in other PRs!

:shipit:

davewhiteland commented:
I'm OK with "strategies" fwiw

chrismytton commented:
> There are still a few things to tidy up here (e.g. I think we need to document better what a basic Scraped::Request is, as currently we introduce it en passant as how you get a Scraped::Response with no real explanation), and it would be good to find a way to remove the duplication of the Class-or-Hash argument shuffle.

Yep, both good points. I've opened #27 for improving the `Scraped::Request` docs and #28 for finding a better alternative to the Class-or-Hash argument shuffle.

> I also suspect we might want a better user-facing name than 'strategies:' but it should be easy enough to do that in a backwardsly compatible manner when we come up with something.

Yep, I don't foresee any issues with staying backwards compatible if we do want to rename them in the future.

> So I think this is definitely good to merge and evolve further in other PRs!

🎉
